### ISSUE TITLE
More matrix-free :)

### ISSUE THREAD

**Issue by dkarrasch** [MKAbdElrahman/LazyFDFD#1]
I came by to see what type of application of `LinearMaps.jl` functionality you have, and I like what I see. 😉 I just wanted to point out some further options to make it even more matrix-free. Particularly, note the `UniformScalingMap` type, which is a lazy representation of scalar multiples of identity maps, unlike `LinearAlgebra.UniformScaling`, however, with a fixed size attached and therefore conveniently usable in `kron` contexts. For instance, instead of `spzeros(nx, nx)` you could use `LinearMap(false*I, nx)`, instead of `Diagonal(I,nz)` you could use `LinearMap(I, nz)`, which reduces storage from the diagonal vector of `nz``Bool` values down to storing a single `Bool` value plus the `Int` for the size `nz`. Similarly, `seye(Nw)` could be `LinearMap(1.0I,Nw)`, unless you need it to be indexable/mutable like an array. Just some suggestions for consideration.

**Comment by dkarrasch** [comment]
x-ref: https://github.com/Jutho/LinearMaps.jl/pull/139

**Comment by MKAbdElrahman** [comment]
Thank you so much for the LinearMaps package. It allowed me to bypass the memory limitations I have and solve very large problems.

I've been trying to benchmark variations between sprase identities, Diagonal(I,N) and LinearMap(I,N) inside Kronecker products. 
Also, the speed of the matrix vector product operation for each case. Because matrix vector products are the backbone of iterative solvers.


```julia output=markdown 
using LinearAlgebra
using SparseArrays
using LinearMaps
using BenchmarkTools


function ∂(N::Int )
	    d = -ones(N)
	    u = ones(N-1)
	    D = spdiagm(0=> d,1=> u)
	    D[N,1] = 1.0
	    return D
end


N = 1000;
∂wf = ∂(N) ;


# Nonlazy implementations could go out of memory before wrapping them !
#@btime kron(Diagonal(I,N), ∂xf, Diagonal(I,N)) # out of memory for large N
#@btime kron(sparse(I,N,N), ∂xf, sparse(I,N,N)) # out of memory for large N


@btime D1 = kron(LinearMap(I,N), ∂wf, LinearMap(I,N))
@btime D2 = kron(Diagonal(I,N), LinearMap(∂wf), Diagonal(I,N))
@btime D3 = kron(LinearMap(sparse(I,N,N)),∂wf, LinearMap(sparse(I,N,N)))
@btime D4 = kron(sparse(I,N,N),LinearMap(∂wf), sparse(I,N,N))

A(D) = [D  D ; D  D] *  [D  D ; D  D]'

function matrix_vector_product(D)
	x = rand(2*size(D,1))
	return A(D) * x
end


@btime matrix_vector_product(D1)
@btime matrix_vector_product(D2)
@btime matrix_vector_product(D3)
@btime matrix_vector_product(D4)

```  


The conclusion from these benchmarks is although D1 has less memory allocations,  D2 has much better performance for matrix vector products. 

We need to balance the saving in memory and the decrease in execution speed. Too much laziness  is like death.

**Comment by dkarrasch** [comment]
That's interesting stuff. First of all, this seems to use pretty much every functionality of LinearMaps.jl. That's awesome.

So, I'd rule out the sparse identities. They cannot be any better than `Diagonal`s. The second observation is that `D1` on its own is faster than `D2`! So the performance flip happens in the concatenation, and I'll need to dig into that to find out why. One reason could be that matrices come with natural 5-arg `mul!`, whereas LinearMaps-`KroneckerMap`s don't. This is used in concatenated maps, so, in principle, it should affect both `D1` and `D2`. Another optimization opportunity might be to make `∂` a `FunctionMap`. Then you don't need to store that simple matrix. It might be favorable, though, to write a little `DiffMap <: LinearMap` struct and provide it with a 5-arg `mul!`. I'll post some suggestions as soon as I get to it.

**Comment by dkarrasch** [comment]
Together with some recent effort (https://github.com/Jutho/LinearMaps.jl/pull/142), you may benefit significantly if you write your own derivative map type like [here](https://gist.github.com/dkarrasch/036d357ee17080277d0eed56bc91f139). You could replace `∂wf = ∂(N)` by `∂wf = DerivativeMap(N)`. You could, in principle, do it via a `FunctionMap`, but the "problem" is that you will want to have fast 5-arg `mul!`. Depending on where your derivative is in the Kronecker product (first, second or third slot), you may even benefit from turning on the `X::AbstractMatrix` methods, but, unfortunately, that's not a generally superior approach. Nevertheless, even without the matrix-methods you get excellent performance, and in case the derivative is in the first position, best performance from everything I've tested.

**Comment by MKAbdElrahman** [comment]
Thank you for providing these great examples. 

The matrixfree `DerivativeMap`  was implemented as you did here  [MatrixFreeMap](https://github.com/MKAbdElrahman/LazyFDFD/tree/master/src/MatrixFreeMap), when I mix these maps with Identities and Zeros of various types as [here](https://github.com/MKAbdElrahman/LazyFDFD/blob/master/src/kron.jl)  I still find wrapping sparse derivatives by `LinearMaps` and using `Diagonal`s  gives the best performance.  `MatrixFreeMaps`  tend to have  large numbers of memory allocations when used in iterative solvers and are slower compared to they wrapped counterparts. 

You can see this here [test3d.jl](https://github.com/MKAbdElrahman/LazyFDFD/blob/master/experiments/test3d.jl)

I noticed also, the iterative solvers behaves differently for sparse arrays and linear maps! . Sparse Arrays have faster speeds per iteration,  but comes with large number of iterations. LinearMaps are slightly slower per iteration, but have an overall smaller number of iterations, so  overall they converge faster.  I don't understand why this happens.

I think for testing, Poisson equation could be  a better test,  since in 3D Helmholtz many things are happening together .  I will test for Poisson in 1D and 2D and comment soon what happens.

**Comment by dkarrasch** [comment]
I noticed that in the DerivativeMap.jl` file, you say that `MulStyle(::DerivativeMap) = FiveArg()`, but then you don't include 5-arg `_unsafe_mul!` methods. I mean those that have two `Number` arguments alpha and beta at the end. Ideally, you should be able to call `mul!(y, D, x, true, true)` without any allocations and superfast. That should also help with rescalings like `DerivativeMap(nl,plane,e⁻ⁱᵏᴸ)/dl`, because the `1/dl` gets moved internally to the alpha argument. I believe this could be one reason for the extra allocations.

As for mixing with matrices, I agree, you should pick what is fasted for you, but from the `LinearMaps.jl` point of view I'd like to get the chance to understand if there is something to be optimized, because these matrices are so trivial. It would be a shame if we couldn't replace them by functions.

**Comment by dkarrasch** [comment]
I'm hooked up to your repo now and did some experiments with the 3dtest file. I noticed you added 5-arg multiplication versions. There

https://github.com/MKAbdElrahman/LazyFDFD/blob/47c0e32140d223c8a60ba7a141428703396419df/src/MatrixFreeMap/DerivativeMap.jl#L31

you need to have
```julia
y[i] = (x[i+1] - x[i])*α + y[i]*β
```
just as elsewhere. If you want to avoid the few extra operations, you could have both 3-arg and 5-arg methods. Further, I noticed in my local tests that the performance issue seems to stem from `∂y`. If you make that `∂y =   ∂s(y,plane,g ,e⁻ⁱᵏᴸ = e⁻ⁱᵏᴸ[2] )` (instead of `∂map`), then you can use safely use `LZero` and `LIdentity` throughout, seemingly without any noticable performance drop. Would be interesting if you could confirm that. And still I'd be interested what exactly causes the slowdown in the `∂y`. BTW, I think it would help if you `up`grade to the latest LinearMaps.jl release.

**Comment by MKAbdElrahman** [comment]
I updated  LinearMaps 

The matrix vector product A*x is  nearly 3.4 times faster when A is a full sparse array.

1-  Converting `∂map` for `∂x`  to `∂s` in the `MatrixFreeCurl` makes A*x 1.5 times faster.  
2-  Converting `∂map` for `∂y`  to `∂s` in the `MatrixFreeCurl` makes A*x 1.2 times faster.  
3-  Converting `∂map` for `∂z`  to `∂s` in the `MatrixFreeCurl` makes A*x 1.2 times faster.  
4-  Converting `∂map` for `∂x` and `∂z`  to `∂s` in the `MatrixFreeCurl` makes A*x   2 times faster.  
5-  Converting `∂map` for `∂x` and `∂y`  to `∂s` in the `MatrixFreeCurl` makes A*x   2 times faster.  
6-  Converting `∂map` for `∂y` and `∂z`  to `∂s` in the `MatrixFreeCurl` makes A*x   1.5  times faster.  
7-  Converting `∂map` for `∂y` and `∂z`  and `∂x`  to `∂s` in the `MatrixFreeCurl` regains  A*x   3.4  times faster.  

So, `∂x` looks to be the most limiting map.    
- `LZero`  even makes the performance better than `DZero`.
- `LIdentity` kills performance compared to `DIdentity` (20 times slower for case 1 above and 7 times slower for case 2).

So, the performance is much better now (than `WrappedSparseCurl`) when I use case 1 and `LZero`.

Could you please explain why do I need the 5-Arg mul! and 3-Arg mul! to both exist ? Thank you :)

**Comment by MKAbdElrahman** [comment]
Although the matrix-vector product A*x is faster for  `A::SparseArray` than for `A::LinearMap`. The iterative solver behaves differently and converges much faster and more stable for `A:LinearMap` than for `A::SparseArray`. This is very strange  as they are the same matrices.

**Comment by dkarrasch** [comment]
> Could you please explain why do I need the 5-Arg mul! and 3-Arg mul! to both exist ? Thank you :)

If you have only one with default values for alpha and beta, but call `mul!(y, A, x)`, then alpha and beta get set to their default values and you have the little extra multiplications and the addition. Having the 5-arg `mul!` however helps a lot in saving memory when you have concatenated maps.

**Comment by MKAbdElrahman** [comment]
I know always `alpha = 1` and `beta  =   0`. would it make a difference if I put them inside the `mul!` function ?.  It adds only the extra computation. I written the 5arg `mul!` as it saves memory as you said and overlooked the extra computation.

**Comment by dkarrasch** [comment]
No, when you have concatenated maps, say horizontally, then you "3-arg mul!" the first map of the row, and then "5-arg mul!" the second map with `alpha=beta=1`, by how it is implemented in `LinearMaps.jl`. You may need it because we need it. 😉  But in case you don't need it in a specific call, you may save a few operations if you also have a pure 3-arg mul! But I don't think that is a bottleneck in your case.

**Comment by MKAbdElrahman** [comment]
Thank you for the intuitive explanation :). Both 5-arg and 3-arg `_unsafe_mul!` are implemenetd now [here](https://github.com/MKAbdElrahman/LazyFDFD/blob/master/src/MatrixFreeMap/DerivativeMap.jl)

**Comment by dkarrasch** [comment]
Alright, changing
```
-- ∂w = DerivativeMap(nl,plane,e⁻ⁱᵏᴸ)/dl
++ ∂w = Ds(plane,nl,e⁻ⁱᵏᴸ)/dl
```
in the SCPML `∂map` makes
```julia
Af =  MatrixFreeCurl(dual,g,pml,ω)*MatrixFreeCurl(primal,g,pml,ω) - ω^2*μ0*ϵ0*I
```
even slightly faster than
```julia
As =  SparseCurl(dual,g,pml,ω)*SparseCurl(primal,g,pml,ω) - ω^2*μ0 *ϵ0*I
```
when using `LZero` and `LIdentity` in `kron.jl`. Apparently, for some stupid reason that I don't understand yet, it is beneficial for both `Af` and `As` to use `DIdentity`, which reduces the time for one `A[*] * J` from about 25 ms down to 1.995 ms (for `Af`) and to 1.121 ms (for `As`). This is a bit disappointing, and it would be interesting to see where the difference comes from. Even more so, because with `LIdentity` the map partial derivatives individually are faster than the sparse partial derivatives. So the concatenation must revert this effect.

**Comment by dkarrasch** [comment]
But I can confirm that the iterative solver is twice as fast for `Af`!

**Comment by MKAbdElrahman** [comment]
sure, the change 
```julia
-- ∂w = DerivativeMap(nl,plane,e⁻ⁱᵏᴸ)/dl
++ ∂w = Ds(plane,nl,e⁻ⁱᵏᴸ)/dl
```

will lead to a faster matrix-vector productucs as fast as  sparse arrays with additional benefit of fast and stable convergence.  But the `O` in the concatenation

```julia
[O    -∂z     ∂y;
∂z     O     -∂x;
-∂y    ∂x      O ]
```
has to be  a of type `LinearMap`  to avoid the explicit memory allocation. 

The problem is using sparse arrays instead of linear maps 
```julia
++ ∂w = Ds(plane,nl,e⁻ⁱᵏᴸ)/dl
```
inside `kron` with `Diagonal(I,n)` will produce non lazy arrays. `LinearMap(I,n)`  would help avoid this issue, but not ready yet  to use. Wrapping `Ds` is not also good.

**Comment by MKAbdElrahman** [comment]
I'd  like to emphasize that using the Lazy zero is crucial here in
```julia 
A = [O    -∂z     ∂y;
∂z     O     -∂x;
-∂y    ∂x      O ]
``` 
A diagonal zero causes a storage space 18  times in  my case as we have two `As` multiplied by each other.

There should be some trade off between speed and memory storage. I suppose a linearmap leads to a code full of dense function calls which could be slower than direct access of array entries.   If there were somwthing like this, how can we quantify that ? So we can tell the cost of using LinearMaps on performance :)

**Comment by dkarrasch** [comment]
Yes, the lazy zero (map) is no problem. The "problem" is the lazy product inside the Kronecker product inside the concatenation inside the product of `MFC(dual)*MFC(primal)`. That specific single lazy product of `∂w` with zeta pretty much spoils much of the performance, together with that strange lazy identity.

> The problem is using sparse arrays instead of linear maps [...] inside kron with Diagonal(I,n) will produce non lazy arrays. 

When you take `kron` and at least one argument is a `LinearMap` (among the first 8, but that's safe in your case), then all factors are wrapped in LinearMaps. Or, when you concatenate maps and arrays, then all arrays are wrapped by LinearMaps as well.

**Comment by MKAbdElrahman** [comment]
To get some performance back, it's better to avoid ScaledMaps and have all the scalars inside the  definition of function maps .  Doing so increases  the speed matrix-vector multiplication by 30% percent
As you can find [in these benchmarks] (https://github.com/MKAbdElrahman/PoissionMap/blob/master/src/test_scaling.jl)

**Comment by dkarrasch** [comment]
> To get some performance back, it's better to avoid `ScaledMap` and have all the scalars inside the definition of function maps . Doing so increases the speed of matrix-vector multiplication by 30% percent
> As you can find [in these benchmarks](https://github.com/MKAbdElrahman/LaplacianMap/blob/master/src/test_scaling.jl)

This is because your Delta doesn't have a 5-arg `_unsafe_mul!`. If it has, `LinearMaps.jl` is going to move the scalar into the alpha argument, and then it's "in the code". That is my general suggestion: if you can, write 5-arg mul! methods and make `MulStyle(::MyMapType) = FiveArg()`. If you don't mind a little "code duplication", write both 3-arg and 5-arg methods, and consider writing methods that work on matrices `mul!(Y::AbstractMatrix, A, X::AbstractMatrix[, alpha, beta])`. This might be helpful in Kronecker products because of the reshape.

> It gives me hope though that **pure** Linear maps performs far better than dense and sparse arrays: [Lazy vs Sparse vs Dense Arrays](https://github.com/MKAbdElrahman/LaplacianMap/blob/master/src/test_dense_vs_lmap.jl).

Yes, that doesn't surprise me at all. I did similar benchmarks when adopting tests from the PyLops paper, see https://github.com/Jutho/LinearMaps.jl/issues/98, and found the same. The reason why I haven't considered including finite-difference operators in LinearMaps.jl is because (i) it's a rather generic package, and (ii) people may want different boundary conditions, central vs one-sided difference and all that, and since it's not too difficult to implement, I thought people should do it themselves, subtype `LinearMap` and have full control.

**Comment by dkarrasch** [comment]
BTW, you can avoid the `TransposeMap{<:Any,<:MyMapType}` in case you have a "transposition-invariant" type.
```julia
Base.transpose(A::Δ) = A # if it's symmetric
```
This can also handle `A'` because I assume your eltype `T` is always real, and for real `A`, `adjoint(A) = transpose(A)`. Another `LinearMaps.jl`-example is, for instance, `CompositeMap`, where the adjoint of a composite map is again a composite map, so you never get a chance to construct a `TransposeMap{<:Any,<:CompositeMap}`.

**Comment by MKAbdElrahman** [comment]
`5-arg _unsafe_mul!` is now [implemented](https://github.com/MKAbdElrahman/LaplacianMap/blob/master/src/laplacian.jl), but the speed gap hasn't changed.  


> BTW, you can avoid the `TransposeMap{<:Any,<:MyMapType}` in case you have a "transposition-invariant" type.

Also changing inearMap default is cleaner: 
```julia
LinearAlgebra.issymmetric(::Δ) = true

```
>

**Comment by dkarrasch** [comment]
> `5-arg _unsafe_mul!` is now [implemented](https://github.com/MKAbdElrahman/LaplacianMap/blob/master/src/laplacian.jl), but the speed gap hasn't changed.

The difference comes from whether the code runs through the 3-arg or the 5-arg `mul!`. Try
```julia
function _unsafe_mul!(y::AbstractVector, A::Δ, x::AbstractVector)
    LinearAlgebra.require_one_based_indexing(x)
    iδ² = inv(A.dx^2)
    @inbounds begin
        y[1] = (x[2] - 2x[1])*iδ²
        for i in 2:A.N-1
            y[i] = (x[i+1] - 2x[i] + x[i-1])*iδ²
        end
            y[end] = (-2x[end] + x[end-1])*iδ²
    end
    return y
end

function _unsafe_mul!(y::AbstractVector, A::Δ, x::AbstractVector,α::Number,β::Number)
    LinearAlgebra.require_one_based_indexing(x)
    iδ²α = inv(A.dx^2)*α  
    if !iszero(β)
        @inbounds begin
            y[1] = (x[2] - 2x[1])*iδ²α + y[1]*β
            for i in 2:A.N-1
                y[i] = (x[i+1] - 2x[i] + x[i-1])*iδ²α + y[i]*β
            end
            y[end] = (-2x[end] + x[end-1])*iδ²α + y[end]*β
        end
    else
        @inbounds begin
            y[1] = (x[2] - 2x[1])*iδ²α
            for i in 2:A.N-1
                y[i] = (x[i+1] - 2x[i] + x[i-1])*iδ²α
            end
            y[end] = (-2x[end] + x[end-1])*iδ²α
        end
    end
    return y
end
```
and then
```julia
@btime mul!($x, $A, $b)
@btime mul!($x, $A, $b, true, false)
@btime mul!($x, $(5A), $b, true, false)
```
then it's even faster than before. Inverting once and multiplying is significantly faster than dividing in every loop.

> Also changing LinearMap default is cleaner:

Yes, indeed! But that's irrelevant for performance. That is important only for (iterative) (eigen)solvers.

**Comment by MKAbdElrahman** [comment]
sounds great!  I looked at the code for some IterativeSolvers and it just call 3-arg `mul!` . From my perspective, its better to force  these solvers to always call 5-Arg `mul!` 
 so, a wrapper like this will work 
``` julia
function _unsafe_mul!(y::AbstractVector, A::Δ, x::AbstractVector)
    _unsafe_mul!(y, A, x,true,false)
end
```
The extra cost is an if condition if beta was `0` an its negligable.

So, `MulStyle` looked a bit confusing to me at start. I thought of a controller to which type of `mul!` is called , but then I found the number of arguments of `mul!` is what controls the dispatching. The confusion comes from that I  call `x = A*b`  and not `mul!(x,A,b)`.  
 `Mulstyle` has to do something with speeding block multiplication when beta = `one`.   The user may have only implemented a 3-arg `mul!` and MulStyle would  do its role. However, if he has a 5-Arg `mul!`  he doesn't need to define `MulStyle` ?

**Comment by dkarrasch** [comment]
> `Mulstyle` has to do something with speeding block multiplication when beta = `one`.

Exactly, or when you have a linear combination, then the first one calls the 3-arg mul! and the rest calls 5-arg with beta=true=1. `MulStyle` tells the package whether you need an intermediate vector to store `A*x` before adding it to `y`, in some cases. Better to have it consistent. 😉 

> From my perspective, its better to force these solvers to always call 5-Arg mul!

I don't think you need that, unless you don't like the "kind of code duplication". But of course, you can do what you suggest in case it makes no performance difference.

**Comment by MKAbdElrahman** [comment]
> I don't think you need that, unless you don't like the "kind of code duplication". But of course, you can do what you suggest in case it makes no performance difference.

I need so,  since ScaledMaps don't perform well  with 3-arg `mul!`.   If I add both definitions I will always suffer performance of ScaledMap because 3-arg `mul!` is the default of A*x at  least

**Comment by dkarrasch** [comment]
You can look at the `LinearMaps.jl` code how ScaledMaps are handled: `mul!(y, 5*A, x) = mul!(y, A, x, 5, false)`. The benchmark above showed that scaled and non-scaled maps are equally fast, at least in isolation. I'd need to see a different benchmark that shows a performance difference to dig into it.

**Comment by MKAbdElrahman** [comment]
Thank you! `mul!(y, 5*A, x) = mul!(y, A, x, 5, false)` and if 5arg-`mul!` is defined will bypass the definition of `3-arg mul!`
as in this  simple example
```julia
f(x,y) ="x+y"
f(x) = f(x,0)
f(x) = "x"
```
calling `f(x)` will print `"x"`

**Comment by dkarrasch** [comment]
> if 5arg-`mul!` is defined will bypass the definition of `3-arg mul!`

Exactly, it's going to bypass 3-arg mul! for the scaled version of `A`, but the unscaled version of `A` may still call the 3-arg mul!.

In your example above, your are overwriting the `f(x) = f(x,0)` method by the `f(x) = "x"` method. After the first method is overwritten, it doesn't exist and cannot work anymore.

**Comment by MKAbdElrahman** [comment]
```julia
       function _unsafe_mul!(y::$Out, A::ScaledMap, x::$In{<:RealOrComplex})
            return _unsafe_mul!(y, A.lmap, x, A.λ, false)
        end
```

I got it.  I defined `3arg _unsafe_mul!` for the type `laplacian` map not the scaled type.    

multiple dispatch will direct  the more specialized scaled type to 5-arg mul! (if it exists) 

if 5-arg `mul!` didn't exist, `ScaledMap` is a subtype of `LinearMap` so it will just call my 3-arg `mul!`

**Comment by dkarrasch** [comment]
> multiple dispatch will direct the more specialized scaled type to 5-arg mul! (if it exists)

Correct. In fact, it will always redirect to the 5-arg `mul!`. In case there is no specialized one, it will be caught by the generic `LinearMap` method as you say, which will further direct it to `_generic_mapvec_mul!` or something like that, with the potential "disadvantage" that it will allocate that intermediate vector and use the 3-arg `mul!` to multiply into it.

**Comment by MKAbdElrahman** [comment]
I noticed that  taking  linear combinations of linear maps slows down  matrix-vector  products significantly. 

```julia
Float = Float64
N = 10 ; dx = 1.2 ;
A = Δ{Float}(N, dx) #A:LinearMap

Id = LinearMap(I,N);
Amap = kron(Id,A) +  kron(Id,A) +  kron(Id,A)
Amap2 = 3*kron(Id,A)

x = rand(N^2); 

@btime Amap*x
@btime Amap2*x

``` 
Matrix vector products slows down inversely with the number of term in the linear combination.

**Comment by dkarrasch** [comment]
Yes, not surprisingly. The `LinearCombination` `kron(Id,A) +  kron(Id,A) +  kron(Id,A)` computes the action of `kron(Id,A)` on a vector `x` three times, because it doesn't know that the summands are the same (and it shouldn't even check for the sake of type stability), whereas `3*kron(Id,A)` computes the action only once and then scales the result by 3. As for the composition, you may save a little by preallocating a vector for the intermediate result if you know it's always going to be a product of two maps.

**Comment by dkarrasch** [comment]
If you happen to know simplifying algebraic rules, you should make use of them yourself. LinearMaps is not a Computer-Algebra-System. If I remember correctly, there was only one product in the MatrixFreeCurl that I suggested to replace which seemed to be the major bottleneck...?

You could write your own `TwoFactorMap` type, which is essentially a `CompositeMap` with two maps and a preallocated work array. Something like this:
```julia
struct TwoFactorMap{T,A<:CompositeMap{T,<:Tuple{LinearMap,LinearMap}}}
    lmap::A
    temp::Vector{T}
end
```
and then hook up its `_unsafe_mul!` to the `_composite_mul!`. It's internal stuff, but at some point we splitted the `CompositeMap` `mul!` function into two parts to be able to provide the temporary vector. But you may as well try to replace some of the compositions by their sparse matrix representation, if that doesn't flood your memory.

**Comment by MKAbdElrahman** [comment]
At least I understand now the source of  slowness. kron of linear maps by themselves are faster than sparse arrays.  

but Looking at the curl:

```julia
O    -∂z     ∂y;
 ∂z     O     -∂x;
 -∂y    ∂x      O
 ]
```
for which I'm going to multiply it by  almost itself again (transpose in many cases). That's large number of compositions!

The other source my be concatenation.  I will experiment with this more.

**Comment by dkarrasch** [comment]
> That's large number of compositions!

No, it's just one composition, as you can see from the display of the map. LinearMaps.jl is not going to multiply out the blocks. It will keep the composition of two BlockMaps. I think one major bottleneck is the product in `kron(ζw * ∂w, axis, g)` in
```julia
function ∂map(axis::Axis,plane::GridPlane,g::Grid{D},pml::SCPML{D},ω::Real;e⁻ⁱᵏᴸ::Number = 1.0) where D
```
It might help to have `∂w = Ds(nl,plane,e⁻ⁱᵏᴸ)/dl` there. But I have no idea where it's feasible to allocate matrices (even if sparse), and where it's not. There may be some price to pay for matrix-freeness.

> The other source my be concatenation.

I think the combination of vector views (from the mul! code for concatenations) and reshapes (of views) (from the mul! code for Kronecker products) is indeed a difficult thing. But you have particularly deep nestings of LinearMaps types, which I find exciting!

**Comment by dkarrasch** [comment]
Cool, so you could have a little "diagonally-scaled" variation of `DerivativeMap`.
